sonic_data_client/mixed_db_client.go

```diff
@@ -75,6 +75,58 @@ type MixedDbClient struct {
 
 var mixedDbClientMap = map[string]MixedDbClient{}
 
+func hget(table swsscommon.Table, key string, field string) (string, error) {
```

Comment:
Add this method because the SWIG can't handle c++ return by reference parameter feature, the generated wrapper for following method will crash:


    virtual bool hget(const std::string &key, const std::string &field,  std::string &value);

fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x7 pc=0x7f978cdeaafb]

runtime stack:
runtime.throw(0x150c215, 0x2a)
	/usr/lib/go-1.15/src/runtime/panic.go:1116 +0x72
runtime.sigpanic()
	/usr/lib/go-1.15/src/runtime/signal_unix.go:726 +0x4ac

goroutine 50 [syscall]:
runtime.cgocall(0x11ecdd0, 0xc00039dba0, 0x4c0100)
	/usr/lib/go-1.15/src/runtime/cgocall.go:133 +0x7a fp=0xc00039db58 sp=0xc00039db20 pc=0x4c0c7a
github.com/sonic-net/sonic-gnmi/swsscommon._Cfunc__wrap_Table_hget_swsscommon_a53b54aefdf37e88(0x34fd550, 0x14e72bc, 0x6, 0x14e78ce, 0x6, 0xc0004f4c40, 0x4c9000)
	_cgo_gotypes.go:13277 +0x68 fp=0xc00039dba0 sp=0xc00039db58 pc=0xc384a8
github.com/sonic-net/sonic-gnmi/swsscommon.SwigcptrTable.Hget.func1(0x34fd550, 0xc00039dc58, 0xc00039dc48, 0xc0004f4c40, 0x10)
	/sonic/src/sonic-gnmi/swsscommon/swsscommon.go:10687 +0x148 fp=0xc00039dc10 sp=0xc00039dba0 pc=0xc6b7a8
github.com/sonic-net/sonic-gnmi/swsscommon.SwigcptrTable.Hget(0x34fd550, 0x14e72bc, 0x6, 0x14e78ce, 0x6, 0xc0004f4c40, 0x1340fc0)
	/sonic/src/sonic-gnmi/swsscommon/swsscommon.go:10687 +0x8c fp=0xc00039dc78 sp=0xc00039dc10 pc=0xc5864c
github.com/sonic-net/sonic-gnmi/swsscommon.(*SwigcptrTable).Hget(0xc0005ba688, 0x14e72bc, 0x6, 0x14e78ce, 0x6, 0xc0004f4c40, 0xc00039dd40)
	<autogenerated>:1 +0x9a fp=0xc00039dcd0 sp=0xc00039dc78 pc=0xc7f57a
github.com/sonic-net/sonic-gnmi/sonic_data_client.getDpuAddress(0x14e5951, 0x4, 0x2a, 0x14e5142, 0x3, 0x14f4205)
	/sonic/src/sonic-gnmi/sonic_data_client/mixed_db_client.go:103 +0x231 fp=0xc00039de10 sp=0xc00039dcd0 pc=0x11c8e51

Reply:
Do you have a link/proof for "SWIG can't handle c++ return by reference parameter feature"? Since this is an application, and similar issue could appear in any other golang application, if we really need to fix it, we should fix it in a central place.

Reply:
In SWIG doc, it's only support output c++ pointer: https://swig.org/Doc4.0/SWIGDocumentation.html#Go_output_arguments

I checked the c++ wrapper generate by SWIG, it's because SWIG convert the go string pointer to a reference and pass to c++, Then c++ modify and break the go string memory.

I don't think we can fix this issue with only change the swsscommon.i, to fix this issue we need add new method to swsscommon.